app/gui2/src/components/GraphEditor/snap.ts

```diff
@@ -0,0 +1,152 @@
+import { partitionPoint } from '@/util/array'
+import type { Opt } from '@/util/opt'
+import { Rect } from '@/util/rect'
+import { Vec2 } from '@/util/vec2'
+import { abs } from 'lib0/math'
+import type { ExprId } from 'shared/yjsModel'
+import { computed, markRaw, type ComputedRef } from 'vue'
+
+export class SnapGrid {
+  leftAxes: number[]
+  rightAxes: number[]
+  topAxes: number[]
+  bottomAxes: number[]
+
+  constructor(rects: Rect[]) {
+    markRaw(this)
+    this.leftAxes = Array.from(rects, (rect) => rect.left).sort((a, b) => a - b)
+    this.rightAxes = Array.from(rects, (rect) => rect.right).sort((a, b) => a - b)
+    this.topAxes = Array.from(rects, (rect) => rect.top).sort((a, b) => a - b)
+    this.bottomAxes = Array.from(rects, (rect) => rect.bottom).sort((a, b) => a - b)
+    console.log(this.leftAxes)
+    console.log(this.rightAxes)
+    console.log(this.topAxes)
+    console.log(this.bottomAxes)
+  }
+
+  snappedMany(rects: Rect[], threshold: number): Vec2 {
+    const minSnap = rects.reduce<[number | null, number | null]>(
+      (minSnap, rect) => {
+        const [xSnap, ySnap] = this.snap(rect, threshold)
+        return [SnapGrid.minSnap(minSnap[0], xSnap), SnapGrid.minSnap(minSnap[1], ySnap)]
+      },
+      [null, null],
+    )
+    return new Vec2(minSnap[0] ?? 0.0, minSnap[1] ?? 0.0)
+  }
+
+  snap(rect: Rect, threshold: number): [number | null, number | null] {
```

Comment:
labeling the entries in the tuple would be nice: `[x: number | null, y: number | null]`
also it looks like the `?? 0.0` means the type of the second element would be simply `number`

Reply:
`?? 0.0` is actually an error, thanks for catching.